dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,263 @@
+
+        const auto packagedProcessHResult = ::GetCurrentPackageFullName(&packageFullNameLength, packageFullName);
+
+        if (packagedProcessHResult == APPMODEL_ERROR_NO_PACKAGE)
+        {
+            return false;
+        }
+
+        THROW_IF_WIN32_ERROR(packagedProcessHResult);
+
+        return true;
+    }
+
+    winrt::IAsyncOperationWithProgress<winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelResult, winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelStatus> PushNotificationManager::CreateChannelAsync(winrt::guid remoteId)
+    {
+        static bool s_remoteIdInProgress;
+        static wil::critical_section s_lock;
+
+        THROW_HR_IF(E_INVALIDARG, (remoteId == winrt::guid()));
+
+        // API supports channel requests only for packaged applications for v0.8 version
+        THROW_HR_IF(E_NOTIMPL, (IsPackagedProcess() == false));
+
+        winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelResult channelResult{ nullptr };
+        {
+            auto lock = s_lock.lock();
+            if (s_remoteIdInProgress == false)
+            {
+                s_remoteIdInProgress = true;
+            }
+            else
+            {
+                channelResult = winrt::make<winrt::Microsoft::Windows::PushNotifications::implementation::PushNotificationCreateChannelResult>(
+                    nullptr, WPN_E_OUTSTANDING_CHANNEL_REQUEST, PushNotificationChannelStatus::CompletedFailure);
+                co_return channelResult;
+            }
+        }
+
+        auto scopeExit = wil::scope_exit([&]()
+            {
+                auto lock = s_lock.lock();
+                s_remoteIdInProgress = false;
+            });
+
+        auto cancellation{ co_await winrt::get_cancellation_token() };
+
+        cancellation.enable_propagation(true);
+
+        // Allow to register the progress and complete handler
+        co_await resume_background();
+
+        auto progress{ co_await winrt::get_progress_token() };
+
+        uint8_t retryCount = 0;
+        winrt::hresult channelRequestResult = E_PENDING;
+        PushNotificationChannelStatus status = PushNotificationChannelStatus::InProgress;
+
+        PushNotificationCreateChannelStatus
+            channelStatus = { channelRequestResult, status, retryCount };
+
+        progress(channelStatus);
+
+        PushNotificationChannelManager channelManager{};
+        winrt::PushNotificationChannel pushChannelReceived{ nullptr };
+
+        for (auto backOffTimeInSeconds = c_minBackoffSeconds; backOffTimeInSeconds <= c_maxBackoffSeconds * 2; backOffTimeInSeconds *= 2)
+        {
+            try
+            {
+                pushChannelReceived = co_await channelManager.CreatePushNotificationChannelForApplicationAsync();
+                channelRequestResult = S_OK;
+                status = PushNotificationChannelStatus::CompletedSuccess;
+                break;
```

Comment:
Just return the answer, instead of setting up a bunch of flags and variables that get tested later. Avoids the need to predeclare stuff, and avoids destructing empty smart pointer variables. Work with the compiler rather than against it.

```suggestion
                co_return winrt::make<PushNotificationCreateChannelResult>(
                    co_await channelManager.CreatePushNotificationChannelForApplicationAsync(),
                    S_OK,
                    PushNotificationChannelStatus::CompletedSuccess);
```